tests/Types.hs

```diff
@@ -89,29 +129,156 @@ u = Unboxed 1
             1#
             1##
 
+c :: Const Int a
+c = Const 1
+
+z :: Ap Maybe Int
+z = Ap (Just 3)
+
+w :: Tree Int
+w = Node 3 [Node 4 [], Node 5 []]
+
+#if MIN_VERSION_base(4,5,0)
+data Unit' = Unit'
+  deriving (Eq, Ord, Show, Typeable)
+$(deriveAll ''Unit')
+
+data Product' a b c d = Product' a b c d
+  deriving (Eq, Ord, Show, Typeable)
+$(deriveAll ''Product')
+
+data Sum' a b = Inl' a | Inr' b
+  deriving (Eq, Ord, Show, Typeable)
+$(deriveAll ''Sum')
+
+newtype Const' a b = Const' a
+  deriving (Eq, Ord, Show, Typeable)
+$(deriveAll ''Const')
+
+newtype Ap' f a = Ap' (f a)
+  deriving (Eq, Ord, Show)
+#if __GLASGOW_HASKELL__ >= 708
+deriving instance Typeable Ap'
+#endif
+$(deriveAll ''Ap')
+
+data Unboxed' a
+   = Unboxed' a
+# if MIN_VERSION_template_haskell(2,11,0)
+             Char#
+# endif
+             Double#
+             Float#
+             Int#
+             Word#
+  deriving (Eq, Ord, Show, Typeable)
+$(deriveAll ''Unboxed')
+
+p' :: Product' Char Int Bool String
+p' = Product' 'a' 1 True "b"
+
+s' :: Sum Char Int
+s' = Inl 'c'
+
+u' :: Unboxed' Int
+u' = Unboxed' 1
+# if MIN_VERSION_template_haskell(2,11,0)
+            '1'#
+# endif
+            1.0##
+            1.0#
+            1#
+            1##
+
+c' :: Const' Int a
+c' = Const' 1
+
+z' :: Ap' Maybe Int
+z' = Ap' (Just 3)
+#endif
+
 pkgKey :: String
 pkgKey = "main"
 
 instance Lift Unit where
     lift = genericLiftWithPkg pkgKey
 #if MIN_VERSION_template_haskell(2,16,0)
-    liftTyped = genericLiftTypedCompat
+    liftTyped = genericLiftTypedCompatWithPkg pkgKey
 #endif
 
 instance (Lift a, Lift b, Lift c, Lift d) => Lift (Product a b c d) where
     lift = genericLiftWithPkg pkgKey
 #if MIN_VERSION_template_haskell(2,16,0)
-    liftTyped = genericLiftTypedCompat
+    liftTyped = genericLiftTypedCompatWithPkg pkgKey
 #endif
 
 instance (Lift a, Lift b) => Lift (Sum a b) where
     lift = genericLiftWithPkg pkgKey
 #if MIN_VERSION_template_haskell(2,16,0)
-    liftTyped = genericLiftTypedCompat
+    liftTyped = genericLiftTypedCompatWithPkg pkgKey
 #endif
 
 instance Lift a => Lift (Unboxed a) where
     lift = genericLiftWithPkg pkgKey
 #if MIN_VERSION_template_haskell(2,16,0)
+    liftTyped = genericLiftTypedCompatWithPkg pkgKey
+#endif
+
+-- This instance shows that we can have parametric polymorphism—we don't need
+-- Typeable instances for @a@ or @b@.
```

Comment:
While this comment is technically true, it's not really relevant here, as `genericLiftWithPkg` has nothing to do with `Typeable`. Since you have the same comment prefacing the `genericLift`-based `Lift` instances, I'd vote to just strike this comment.

Reply:
> While this comment is technically true, it's not really relevant here, as `genericLiftWithPkg` has nothing to do with `Typeable`. Since you have the same comment prefacing the `genericLift`-based `Lift` instances, I'd vote to just strike this comment.

Yeah, that was a bit of copy pasta. Gone now.